### PR TITLE
Coral-Spark: Migrate 'CAST(ROW: RECORD_TYPE)' and 'extract_union' transformations from RelNode to SqlNode layer

### PR DESCRIPTION
**Pull request by ljfgem**
Similar to #351, this PR migrates `CAST(ROW: RECORD_TYPE)` and `extract_union` transformations from `RelNode` layer in `IRRelToSparkRelTransformer` to `SqlNode` layer. 

In order to avoid creating a PR that is too big, I will open several more PRs to finish migrating all the transformations in `IRRelToSparkRelTransformer` to `SqlNode` layer and remove `IRRelToSparkRelTransformer`.

Tests:
1. Existing unit tests, which already cover the changes
2. Regression test
3. Querying several complex production views on gateway, no regression

### CODE REVIEW

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    List<SqlNode> newOperands = new ArrayList<>();
+    final SqlCall rowCall = (SqlCall) sqlCall.getOperandList().get(0); // like `ROW(123, 'xyz')` in above example
+    final SqlRowTypeSpec sqlRowTypeSpec = (SqlRowTypeSpec) sqlCall.getOperandList().get(1); // like `ROW(`abc` INTEGER, `def` CHAR(3) CHARACTER SET `ISO-8859-1`))` in above example
+    for (int i = 0; i < rowCall.getOperandList().size(); ++i) {
+      final String fieldName =
+          ((SqlRowTypeNameSpec) sqlRowTypeSpec.getTypeNameSpec()).getFieldNames().get(i).names.get(0);
+      newOperands.add(new SqlIdentifier("'" + fieldName + "'", SqlParserPos.ZERO)); // need to single-quote the field name
+      newOperands.add(rowCall.getOperandList().get(i));
+    }
+    return HiveNamedStructFunction.NAMED_STRUCT.createCall(sqlCall.getParserPosition(), newOperands);
```
Comment (aastha25) [inline coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:51]: Can we rename `HiveNamedStructFunction` -> `CoralNamedStructFunction` and add documentation on CoralNamedStructFunction about what this function represents (without tying it explicitly with hive)

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/ExtractUnionFunctionTransformer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import com.linkedin.coral.spark.containers.SparkUDFInfo;
+
+
+/**
+ * This transformer transforms `extract_union` to `coalesce_struct`.
+ * Instead of leaving `extract_union` visible to Spark, since we adopted the new exploded struct schema (a.k.a struct_tr)
+ * that is different from extract_union's output (a.k.a struct_ex) to interpret union in Coral IR,
+ * we need to swap the reference of `extract_union` to a new UDF that is coalescing the difference between
+ * struct_tr and struct_ex.
+ * See {@link CoalesceStructUtility#COALESCE_STRUCT_FUNCTION_RETURN_STRATEGY} and its comments for more details.
```
Comment (aastha25) [inline coral-spark/src/main/java/com/linkedin/coral/spark/transformers/ExtractUnionFunctionTransformer.java:30]: the documentation is hard to follow, plus the documentation on `CoalesceStructUtility` suggests `coalesce the {@link RelDataType} of struct between Trino's representation and hive's extract_union UDF's representation`. No mention of hive -> spark transformation. 

Implementation of this transformer looks good. Could you take another stab at documentation? maybe add an example like `CastToNamedStructTransformer`

**Review by aastha25 — COMMENTED** [review]
thanks @ljfgem  for the PR! 
Left minor comments around documentation.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 1 above)
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:51]: I would still prefer to call it `HiveNamedStructFunction` because it's only used in `HiveSqlNode` and `SparkSqlNode`, and it's a Hive/Spark specific function.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.calcite.sql.SqlRowTypeNameSpec;
+import org.apache.calcite.sql.SqlRowTypeSpec;
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import com.linkedin.coral.common.transformers.SqlCallTransformer;
+import com.linkedin.coral.hive.hive2rel.functions.HiveNamedStructFunction;
+
+
+/**
+ * This transformer transforms `CAST(ROW: RECORD_TYPE)` to `named_struct` in Spark.
```
[the remark this answers is not in the document]
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:24]: Not sure if it's suitable to call it `Coral IR` function, I think it's a function in CoralSqlNode converted from Coral IR?

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/ExtractUnionFunctionTransformer.java (hunk 2 above)
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/ExtractUnionFunctionTransformer.java:30]: Thanks for suggestion, added `Spark` to `CoalesceStructUtility` java doc.
As for the example, it mentioned that `Check `CoralSparkTest#testUnionExtractUDF` for examples.`

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aastha25) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:24]: in the path:
`source SQL -> sourceSqlNode1 -> CoralSqlNode1 -> CoralRelNode -> CoralSqlNode2 -> TargetLangSqlNode -> target SQL`

if an operator is present in CoralSqlNode1, CoralRelNode, CoralSqlNode2 and we transform this operator when we go to LanguageSqlNode, such operators can be called Coral IR functions.

[the remark this answers is not in the document]
Comment (aastha25) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:24]: as per the current status of our code base - 
`hive SQL -> HiveSqlNode (same as CoralSqlNode1) -> CoralRelNode -> CoralSqlNode2 -> SparkSqlNode -> Spark SQL`
I think this operator qualifies as Coral Operator

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 1 above)
Comment (aastha25) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:51]: as per current structure, looks like it is used in HiveSqlNode (CoralSqlNode1)  & CoralSqlNode2.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:24]: Thanks for clarification, do you mean an operator is a Coral IR function if it appears in `CoralSqlNode1 / CoralRelNode / CoralSqlNode2` or `CoralSqlNode1 & CoralRelNode & CoralSqlNode2`?
Not sure because it doesn't appear in `CoralSqlNode1`.

[the remark this answers is not in the document]
Comment (aastha25) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java]: for a corresponding input hive SQL, what is the operator we use in the SqlNode representation? how do we represent the named_struct on LHS?

but generally yes, an operator present in CoralSqlNode1 would be present in CoralRelNode and then generated back in CoralSqlNode2.

**Comment by ljfgem** [comment]
@aastha25 Thanks for your review! Modified these 2 comments based on your suggestions. PTAL.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java]: It's still represented as `named_struct` in CoralSqlNode1, and Calcite converts it to `CAST(ROW: RECORD_TYPE)` in RelNode.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 1 above)
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java:51]: It's not used in `CoralSqlNode2`, it's only used in HiveSqlNode and SparkSqlNode.

File: coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aastha25) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java]: I see. In that case, could you investigate if it's possible to add a lightweight override in `CoralRelNodeToCoralSqlNode` to such that the coralSqlNode2 generated uses the same operator `named_struct` as coralSqlNode1?

[the remark this answers is not in the document]
Comment (ljfgem) [reply coral-spark/src/main/java/com/linkedin/coral/spark/transformers/CastToNamedStructTransformer.java]: Opened #357 for tracking.
